### PR TITLE
ADCM-6125 Configs management

### PR DESCRIPTION
**Pull request by Sealwing**

### CODE REVIEW

File: adcm_aio_client/core/config/_base.py [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+
+
+# FOREIGN SECTION
+#
+# these functions are heavily inspired by configuration rework in ADCM (ADCM-6034)
+
+ROOT_PREFIX = "/"
+
+
+def set_nested_config_value[T](config: dict[str, Any], level_names: LevelNames, value: T) -> T:
```
Comment (a-alferov) [inline adcm_aio_client/core/config/_base.py:312]: Do you think we should somehow highlight functions that change objects on the fly?

File: tests/unit/test_config.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    # inner type won't be checked (list),
+    # but here we pretend "to be 100% sure" it's `list`, not `None`
+    config["root_list", ValueWrapper].set([*config["root_list", ValueWrapper, list].value, new_config["root_list"][-1]])
+
+    root_dict = config["root_dict"]
+    assert isinstance(root_dict, ValueWrapper)
+    assert isinstance(root_dict.value, dict)
+    root_dict.set(None)
+    assert root_dict.value is None
+    assert config["root_dict", ValueWrapper].value is None
```
Comment (a-alferov) [inline tests/unit/test_config.py:68]: 🦆

File: adcm_aio_client/core/config/_base.py [hunk 3]
```diff
@@ -0,0 +1,346 @@
+from abc import abstractmethod
```
Comment (kuhella) [inline adcm_aio_client/core/config/_base.py:1]: Do we plan to add License?

File: tests/unit/test_config.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    config_for_save = config.to_payload()
+    assert config_for_save == data
+
+    # Edit "root" values
+
+    config["root_int", ValueWrapper].set(new_config["root_int"])
+
+    # inner type won't be checked (list),
+    # but here we pretend "to be 100% sure" it's `list`, not `None`
+    config["root_list", ValueWrapper].set([*config["root_list", ValueWrapper, list].value, new_config["root_list"][-1]])
```
Comment (kuhella) [inline tests/unit/test_config.py:61]: also seems strange

File: adcm_aio_client/core/config/_base.py [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+        )
+
+        self._initial_data = data
+        self._json_fields: set[tuple[ParameterName, ...]] = set()
+        self._convert_payload_formated_json_fields_inplace(self._spec, self._initial_data["config"], prefix=())
+
+        self._changed_data = None
+
+    def to_payload(self: Self) -> dict:
+        payload = deepcopy(self._current_data)
```
Comment (kuhella) [inline adcm_aio_client/core/config/_base.py:234]: purpose of this method (I see that it copy)? what problem solve this method?

File: tests/unit/test_config.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    param = config["root_str"]
+    assert isinstance(param, ValueWrapper)
+    assert param.value is None
+
+    param.set("newstring")
+    assert isinstance(config["root_str", ValueWrapper].value, str)
+
+    # Check all values are changed
+
+    config_for_save = config.to_payload()
```
Comment (kuhella) [inline tests/unit/test_config.py:111]: what is motivation

File: tests/unit/test_config.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+    assert isinstance(activatable_group, ActivatableGroupWrapper)
+    activatable_group.activate()
+
+    # Edit JSON field
+
+    # change value separately and set
+    json_field = main_group["inner_json"]
+    assert isinstance(json_field, ValueWrapper)
+    assert isinstance(json_field.value, dict)
+    new_value = deepcopy(json_field.value)
```
Comment (kuhella) [inline tests/unit/test_config.py:91]: why we use deepcopy ? seems should be new_value = json_field.value

File: adcm_aio_client/core/config/_base.py (hunk 3 above)
Comment (Sealwing) [reply adcm_aio_client/core/config/_base.py:1]: Yes, as I know, but it should be done in one place with licence checkring, etc.

File: tests/unit/test_config.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+    # deepcopy, because we want our "source" to be intact
+    # and unchanged by json formating
+    config = EditableConfig(data=deepcopy(data), spec=schema)
+
+    config_for_save = config.to_payload()
+    assert config_for_save == data
+
+    # Edit "root" values
+
+    config["root_int", ValueWrapper].set(new_config["root_int"])
```
[the remark this answers is not in the document]
Comment (Sealwing) [reply tests/unit/test_config.py:57]: user can do this, it's for type checking purposes

I'd make it more clear later

File: adcm_aio_client/core/config/_base.py (hunk 5 above)
Comment (Sealwing) [reply adcm_aio_client/core/config/_base.py:234]: protection of internals from change mostly

**Pull request closed without merging**